File: php/139-word-break.php
class Solution {

    /**
    * @param String $s
    * @param String[] $wordDict
    * @return Boolean
    */
    function wordBreak($s, $wordDict) {
        $sLen = strlen($s);
        $dp = array_fill(0, $sLen + 1, false);
        $dp[$sLen] = true;
        for ($i = $sLen - 1; $i >= 0; $i--) {
            foreach ($wordDict as $w) {
                $wLen = strLen($w);
                if ($i + $wLen <= $sLen && $w === substr($s, $i, $wLen)) {
                    $dp[$i] = $dp[$i + $wLen];
                }
                if ($dp[$i]) {
                    break;
                }
            }
        }
        return $dp[0];
    }
}

File: golang/139-word-break.go
package golang

func wordBreak(s string, wordDict []string) bool {
    dp := make([]bool, len(s) + 1)
    dp[len(dp) - 1] = true
    for i := len(s) - 1; i >= 0; i-- {
        for j := 0; j < len(wordDict); j++ {
            if i + len(wordDict[j]) <= len(s) && wordDict[j] == s[i:i + len(wordDict[j])] {
                dp[i] = dp[i + len(wordDict[j])]
            }
            if dp[i] {
                break
            }
        }
    }
    return dp[0]
}